extensions/mqtt/processors/PublishMQTT.h

```diff
@@ -62,72 +68,116 @@ class PublishMQTT : public processors::AbstractMQTTProcessor {
 
   ADD_COMMON_VIRTUAL_FUNCTIONS_FOR_PROCESSORS
 
-  class ReadCallback {
+  void readProperties(const std::shared_ptr<core::ProcessContext>& context) override;
+  void onTriggerImpl(const std::shared_ptr<core::ProcessContext>& context, const std::shared_ptr<core::ProcessSession>& session) override;
+  void initialize() override;
+
+ private:
+  /**
+   * Counts unacknowledged QoS 1 and QoS 2 messages to respect broker's Receive Maximum
+   */
+  class InFlightMessageCounter {
    public:
-    ReadCallback(PublishMQTT* processor, uint64_t flow_size, uint64_t max_seg_size, std::string topic, MQTTAsync client, int qos, bool retain)
-        : processor_(processor),
-          flow_size_(flow_size),
-          max_seg_size_(max_seg_size),
-          topic_(std::move(topic)),
-          client_(client),
-          qos_(qos),
-          retain_(retain) {
+    void setMqttVersion(const MqttVersions mqtt_version) {
+      mqtt_version_ = mqtt_version;
+    }
+
+    void setQoS(const MqttQoS qos) {
+      qos_ = qos;
+    }
+
+    void setMax(const uint16_t new_limit) {
+      limit_ = new_limit;
     }
 
-    int64_t operator()(const std::shared_ptr<io::InputStream>& stream);
+    // increase on sending, wait if limit is reached
+    void increase();
```

Comment:
hm, good point

any idea how long a timeout we need?  is eg. a hard-coded 1 minute OK, or do we need a new property for this?

Reply:
I'm not sure whether this failure scenario is possible. I'd set it to a shorter period (maybe 5 sec?), and let the processor yield if it can't acquire the semaphore. It could try again on the next trigger.

Reply:
I have added a 5 second timeout in 58bc0f6eaf985051820e0d5f6ae212bb0364db04, but if the timeout expires without a new slot opening up, then the flow file will be routed to failure, because that's what we were doing in the case of other error conditions -- is that OK?

Reply:
I think a rollback would be better, since slow or unstable network conditions combined with a fast trigger interval could easily trigger this condition, and it's not necessarily an error.

Reply:
changed to a rollback in 75e067f4b4d1fbcce7aa7454a7a6c71d466bae32